tools/autograd/templates/VariableType.cpp

```diff
@@ -173,12 +173,61 @@ Variable VariableType::as_variable(const Scalar & scalar) const {
   return make_variable(std::move(tensor));
 }
 
-static void check_inplace(const VariableImpl& pImpl) {
-  if (pImpl.requires_grad && !pImpl.grad_fn) {
+struct VariableFlags {
+  bool requires_grad;
+  bool is_volatile;
+};
+
+template<typename T>
+static VariableFlags compute_flags_tmpl(T tensors) {
+  VariableFlags flags = {false, false};
+  for (const Tensor& tensor : tensors) {
+    auto& var = static_cast<const Variable&>(tensor);
+    if (var.defined()) {
+      flags.requires_grad |= var.requires_grad();
+      flags.is_volatile |= var.is_volatile();
+    }
+  }
+  flags.requires_grad &= !flags.is_volatile;
+  return flags;
+}
+
+using TensorRef = std::reference_wrapper<const Tensor>;
+using TensorRefList = std::initializer_list<TensorRef>;
+
+static VariableFlags compute_flags(const TensorRefList& tensors) {
+  return compute_flags_tmpl(tensors);
+}
+
+static VariableFlags compute_flags(TensorList tensors) {
+  return compute_flags_tmpl(tensors);
+}
+
+static void check_no_requires_grad(const Tensor& tensor, const char* name) {
+  auto& var = static_cast<const Variable&>(tensor);
+  if (var.defined() && var.requires_grad()) {
+    std::string msg = "the derivative for '";
+    msg += name;
+    msg += "' is not implemented";
```

Comment:
Is this the same as the behavior previously, or is this a BC break?

Reply:
We had some, but not all, of these checks in the Python autograd functions.